### PR TITLE
[CH-191] Support generate exec

### PR DESCRIPTION
**Pull request by taiyang-li**
## What changes were proposed in this pull request?

Support generate exec. close https://github.com/Kyligence/ClickHouse/issues/191 

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

test by CH[[194]]

### CODE REVIEW

**Comment by taiyang-li** [comment]
spark plan: 
```
0: jdbc:hive2://localhost:10000/> explain formatted  select explode(array(id, id+1)) as elem, cast(id as string) as spark  from range(3);   


+----------------------------------------------------+
|                        plan                        |
+----------------------------------------------------+
| == Physical Plan ==
CHNativeColumnarToRow (6)
+- * ProjectExecTransformer (4)
   +- * GenerateExecTransformer (3)
      +- RowToCHNativeColumnar (2)
         +- * Range (1)


(1) Range [codegen id : 1]
Output [1]: [id#121L]
Arguments: Range (0, 3, step=1, splits=None)

(2) RowToCHNativeColumnar
Input [1]: [id#121L]

(3) GenerateExecTransformer
Input [1]: [id#121L]
Arguments: explode(array(id#121L, (id#121L + 1))), [id#121L], false, [elem#122L]

(4) ProjectExecTransformer
Input [2]: [id#121L, elem#122L]
Arguments: [elem#122L, cast(id#121L as string) AS spark#115]

(5) WholeStageCodegenTransformer (8)
Input [2]: [elem#122L, spark#115]

(6) CHNativeColumnarToRow
Input [2]: [elem#122L, spark#115]
```

Return result:
```
0: jdbc:hive2://localhost:10000/> select explode(array(id, id+1)) as elem, cast(id as string) as spark  from range(3);   
+-------+--------+
| elem  | spark  |
+-------+--------+
| 0     | 0      |
| 1     | 0      |
| 1     | 1      |
| 2     | 1      |
| 2     | 2      |
| 3     | 2      |
+-------+--------+
```

**Comment by taiyang-li** [comment]
Native arrayJoin in clickhouse backend is almost 3x faster than Spark explode.

```
echo "SELECT * FROM generateRandom('a Array(String), b String', 1, 20, 10) limit 100000 FORMAT Parquet"  | clickhouse-client > test_generate_exec.parquet  

> CREATE TEMPORARY VIEW test_generate_exec
USING org.apache.spark.sql.parquet
OPTIONS (
  path  "/data1/liyang/cppproject/gluten/test_generate_exec.parquet"
) ;

> set spark.gluten.sql.columnar.generate=true;
> select count(*) from (select explode(a), b from test_generate_exec);
+-----------+
| count(1)  |
+-----------+
| 500257    |
+-----------+
1 row selected (0.205 seconds)



> set spark.gluten.sql.columnar.generate=false;
> select count(*) from (select explode(a), b from test_generate_exec);  
+-----------+
| count(1)  |
+-----------+
| 500257    |
+-----------+
1 row selected (0.594 seconds)
 
```

**Comment by taiyang-li** [comment]
@rui-mo Can you help review this pr, thanks!

File: gluten-core/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+import io.glutenproject.expression.ExpressionConverter
+import io.glutenproject.substrait.expression.ExpressionBuilder
+import io.glutenproject.expression.ExpressionTransformer
+import io.glutenproject.substrait.plan.PlanBuilder
+import io.glutenproject.GlutenConfig
+import io.glutenproject.vectorized.ExpressionEvaluator
+import io.glutenproject.backendsapi.BackendsApiManager
+import com.google.common.collect.Lists
+
+case class GenerateExecTransformer(
```
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/execution/GenerateExecTransformer.scala:48]: Could you add some notes for this transformer to describe its functionality and usage?

**Comment by taiyang-li** [comment]
@rui-mo  It it possible running gluten CI/CD with clickhouse version binding to https://github.com/Kyligence/ClickHouse/pull/194 ?

**Comment by rui-mo** [comment]
> @rui-mo It it possible running gluten CI/CD with clickhouse version binding to [Kyligence/ClickHouse#194](https://github.com/Kyligence/ClickHouse/pull/194) ?

Sure. I think it is OK.

File: gluten-core/src/main/scala/io/glutenproject/expression/ArrayExpressionTransformer.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  override def doTransform(args: java.lang.Object): ExpressionNode = {
+    val childNodes = new java.util.ArrayList[ExpressionNode]()
+    children.foreach(child => {
+      val childNode = child.doTransform(args)
+      childNodes.add(childNode)
+    })
+
+    // Ignore useStringTypeWhenEmpty for clickhouse backend
+    val functionMap = args.asInstanceOf[java.util.HashMap[String, java.lang.Long]]
+    val functionName = ConverterUtils.makeFuncName(ExpressionMappings.CREATE_ARRAY,
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/ArrayExpressionTransformer.scala:44]: use `substraitExprName: String` to pass substrait expression name, dont use `ExpressionMappings.CREATE_ARRAY` directly, please refer to `HashExpressionTransformer`

File: gluten-core/src/main/scala/io/glutenproject/expression/ArrayExpressionTransformer.scala [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    with Logging {
+
+  override def doTransform(args: java.lang.Object): ExpressionNode = {
+    val childNodes = new java.util.ArrayList[ExpressionNode]()
+    children.foreach(child => {
+      val childNode = child.doTransform(args)
+      childNodes.add(childNode)
+    })
+
+    // Ignore useStringTypeWhenEmpty for clickhouse backend
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/ArrayExpressionTransformer.scala:42]: does Velox backend need? @rui-mo

File: gluten-core/src/main/scala/io/glutenproject/expression/UnaryExpressionTransformer.scala [hunk 4]
```diff
@@ -65,6 +66,30 @@ class NormalizeNaNAndZeroTransformer(
   }
 }
 
+class ExplodeTransformer(child: ExpressionTransformer, original: Explode)
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/expression/UnaryExpressionTransformer.scala:69]: not need to specified new a transformer for `Explode`, right ?

File: gluten-core/src/main/java/io/glutenproject/substrait/type/ListNode.java [hunk 5]
```diff
@@ -12,6 +12,10 @@ public class ListNode implements TypeNode, Serializable {
     this.nestedType = nestedType;
   }
 
+  public TypeNode getNestedType() {
+    return nestedType;
```
Comment (rui-mo) [inline gluten-core/src/main/java/io/glutenproject/substrait/type/ListNode.java:16]: Can we add some notes here to explain why nested type is needed?

File: gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto [hunk 6]
```diff
@@ -163,6 +163,8 @@ message ExpandRel {
   substrait.extensions.AdvancedExtension advanced_extension = 10;
 }
 
+
+
```
Comment (rui-mo) [inline gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto:167]: Looks one empty line is enough here.

File: gluten-core/src/main/scala/io/glutenproject/expression/ArrayExpressionTransformer.scala (hunk 3 above)
Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/expression/ArrayExpressionTransformer.scala:42]: I find in Spark this config determines the element type to be NullType or StringType when args are empty. In Velox, the NullType way is used by default, but I think the backend needs to follow what Spark plans. Maybe we can send this config to native and mark the function option as `req`.

Comment (taiyang-li) [reply gluten-core/src/main/scala/io/glutenproject/expression/ArrayExpressionTransformer.scala:42]: ClickHouse has the same behavior with Velox.  
```
:) select array() as x,  toTypeName(x);

SELECT
    [] AS x,
    toTypeName(x)

Query id: 73ec966f-00a6-49b7-9936-77e44ce17f14

┌─x──┬─toTypeName(array())─┐
│ [] │ Array(Nothing)      │
└────┴─────────────────────┘

1 rows in set. Elapsed: 0.002 sec. 
```

Comment (taiyang-li) [reply gluten-core/src/main/scala/io/glutenproject/expression/ArrayExpressionTransformer.scala:42]: @rui-mo I think currently we can partially support transforming to substrait plan only when `useStringTypeWhenEmpty` is false, before Velox and ClickHouse supports `useStringTypeWhenEmpty`,

File: gluten-core/src/main/scala/io/glutenproject/expression/UnaryExpressionTransformer.scala (hunk 4 above)
Comment (taiyang-li) [reply gluten-core/src/main/scala/io/glutenproject/expression/UnaryExpressionTransformer.scala:69]: Yes, we need. It's logic is different from common implementations.
- Only support array type as input. 
- Output datetype must be nested type from child expression.

File: gluten-core/src/main/scala/io/glutenproject/expression/ArrayExpressionTransformer.scala (hunk 3 above)
Comment (rui-mo) [reply gluten-core/src/main/scala/io/glutenproject/expression/ArrayExpressionTransformer.scala:42]: Sounds good to me.

**Review by rui-mo — APPROVED** [review]

File: backends-clickhouse/src/main/scala/io/glutenproject/utils/CHExpressionUtil.scala [hunk 7]
```diff
@@ -19,12 +19,10 @@ package io.glutenproject.utils
 import io.glutenproject.expression.ExpressionMappings._
 
 object CHExpressionUtil {
-
```
Comment (zzcclp) [inline backends-clickhouse/src/main/scala/io/glutenproject/utils/CHExpressionUtil.scala:22]: why remove?

**Review by lviiii — APPROVED** [review]
LGTM

**Review by zzcclp — APPROVED** [review]
LGTM

File: gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto [hunk 8]
```diff
@@ -890,6 +892,17 @@ message Expression {
   }
 }
 
+message GenerateRel {
+  RelCommon common = 1;
+  Rel input = 2;
+  
+  Expression generator = 3;
```
Comment (EpsilonPrime) [inline gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto:899]: Should this relation be removed in favor of the official Substrait repo's ExpandRel?

Comment (taiyang-li) [reply gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto:899]: GenerateRel and ExpandRel is different. The former is used to represent spark `lateral view explode/posexpode`, and the latter is used to represent spark `grouping sets` or `with cube`

**Pull request merged**